In [1]:
en_sent="From hill to hill no bird in flight."

zh_sent="千山鸟飞绝。"

print("英文:",en_sent)
print("中文:",zh_sent)

英文: From hill to hill no bird in flight.
中文: 千山鸟飞绝。


In [2]:
def mock_dependency_distance(text):
    words = text.split()
    if len(words) <= 1:
        return 0
    total_dist = 0
    for i in range(len(words) - 1):
        total_dist += 1   # 模拟每个相邻词之间的距离为1
    return total_dist / (len(words) - 1)

en_sent = "From hill to hill no bird in flight."
zh_sent = "千山鸟飞绝。"

mock_mdd_en = mock_dependency_distance(en_sent)
mock_mdd_zh = mock_dependency_distance(zh_sent)

print(f"英文模拟MDD: {mock_mdd_en}")
print(f"中文模拟MDD: {mock_mdd_zh}")

英文模拟MDD: 1.0
中文模拟MDD: 0


In [10]:
import spacy

# 加载英文和中文模型
nlp_en = spacy.load("en_core_web_sm")
nlp_zh = spacy.load("zh_core_web_sm")

print("模型加载成功！")

模型加载成功！


In [12]:
def calculate_mdd(text, nlp):
    doc = nlp(text)
    total_dist = 0
    count = 0
    for token in doc:
        if token.dep_ != "ROOT":
            dist = abs(token.i - token.head.i)
            total_dist += dist
            count += 1
    return total_dist / count if count > 0 else 0

en_sent = "From hill to hill no bird in flight."
zh_sent = "千山鸟飞绝。"

mdd_en = calculate_mdd(en_sent, nlp_en)
mdd_zh = calculate_mdd(zh_sent, nlp_zh)

print(f"真实英文 MDD: {mdd_en:.2f}")
print(f"真实中文 MDD: {mdd_zh:.2f}")

真实英文 MDD: 1.88
真实中文 MDD: 1.00


In [13]:
def calculate_hf(text, nlp):
    doc = nlp(text)
    hf = 0
    total = 0
    for token in doc:
        if token.dep_ == "ROOT":
            continue
        if token.head.i > token.i:
            hf += 1
        total += 1
    return hf / total if total > 0 else 0

hf_en = calculate_hf(en_sent, nlp_en)
hf_zh = calculate_hf(zh_sent, nlp_zh)

print(f"英文 HF: {hf_en:.2f}")
print(f"中文 HF: {hf_zh:.2f}")

英文 HF: 0.38
中文 HF: 0.50


 ## 依存方向 (HF) 计算结果与讨论

对《江雪》首句“From hill to hill no bird in flight”及其译文“千山鸟飞绝”进行依存方向分析：

- 英文 HF = 0.38，说明该英文句子中 head-initial 关系占优（支配词在前），符合英语典型语序。
- 中文 HF = 0.50，显示头尾均衡。理论上汉语更偏 head-final，但本例因句子极短且结构特殊，未表现出明显 head-final 倾向。

**方法验证**：代码成功调用 spaCy 提取依存关系并计算 HF，证明已掌握论文中的核心指标计算流程。后续需使用更大规模平行语料进行统计检验，才能得出可靠结论。